### DISCUSSION TITLE
Issue adding SofaPython3 plugin in runSofa v24.06.00: init_fs_encoding error

### DISCUSSION THREAD

**Discussion by sebastienfricker** [sofa-framework/sofa#5141] [Build / Config / Environment] [answered]
Hello Sofa community,
I am a new user wishing to try Sofa with python.

I installed Sofa v24.06.00 from binaries, on Windows, and runSofa works fine.
I installed python 3.10.15 (in a CONDA environment)

I have also set the system variables as follows:
SOFA_ROOT C:\Users\FRICKERS\SOFA\v24.06.00
PYTHON_ROOT C:\Users\FRICKERS\AppData\Local\anaconda3\envs\SOFA
PYTHONPATH %SOFA_ROOT%\plugins\SofaPython3\lib\python3\site-packages
PATH ...;%PYTHON_ROOT%;%PYTHON_ROOT%\DLLs;%PYTHON_ROOT%\Lib;%SOFA_ROOT%\bin

When I launch runSofa and try to add the SofaPython3 plugin via the plugin manager, I get the following message in the runSofa console:

[INFO]    [SofaPython3] Initializing with python version 3.10.15 | packaged by Anaconda, Inc. | (main, Oct  3 2024, 07:22:19) [MSC v.1929 64 bit (AMD64)]
[INFO]    [SofaPython3] Registering a scene loader for [.py, .py3, .pyscn, .py3scn] files.
[INFO]    [SofaPython3] Initializing python
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = 'C:\Users\FRICKERS\SOFA\v24.06.00\plugins\SofaPython3\lib\python3\site-packages'
  program name = 'python'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = 'C:\\Users\\FRICKERS\\SOFA\\v24.06.00\\bin\\runSofa.exe'
  sys.base_prefix = 'C:\\Users\\FRICKERS\\AppData\\Local\\anaconda3\\envs\\SOFA'
  sys.base_exec_prefix = 'C:\\Users\\FRICKERS\\AppData\\Local\\anaconda3\\envs\\SOFA'
  sys.platlibdir = 'lib'
  sys.executable = 'C:\\Users\\FRICKERS\\SOFA\\v24.06.00\\bin\\runSofa.exe'
  sys.prefix = 'C:\\Users\\FRICKERS\\AppData\\Local\\anaconda3\\envs\\SOFA'
  sys.exec_prefix = 'C:\\Users\\FRICKERS\\AppData\\Local\\anaconda3\\envs\\SOFA'
  sys.path = [
    'C:\\Users\\FRICKERS\\SOFA\\v24.06.00\\plugins\\SofaPython3\\lib\\python3\\site-packages',
    'C:\\Users\\FRICKERS\\AppData\\Local\\anaconda3\\envs\\SOFA\\python310.zip',
    'C:\\Users\\FRICKERS\\SOFA\\v24.06.00\\bin',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Do you have any idea as to what this init_fs_encoding issue could be related ?

**Comment by damienmarchal** [comment]
Hi @sebastienfricker,

I'm not using Windows, but in general python problem like the one you are reporting are because there are interference between different python version (the one expected by sofa and the ones installed in your system). 

From the error message I have the impression that python does not find the default modules ("encodings" is one of the default packages, and thus it seems python is not able to load it). 

Maybe conda is using PYTHONPATH and you setting it on your side thus lossing where conda installs the defaults modules. 

I thinks this may be related to that one:
https://stackoverflow.com/questions/65184937/fatal-python-error-init-fs-encoding-failed-to-get-the-python-codec-of-the-file

One of the answers there is: 
```
I see your PYTHONHOME is set to PYTHONHOME = '/home/env3/educ'. Try to check if it is really there.
The solution for me was to remove the PYTHONHOME environment variable. For you, it can be just that, or setting that variable to another value. This worked on Windows, and would work on Linux for sure. If someone tries this on Linux, please post a comm
```

**Reply by sebastienfricker** [reply]
Hi @damienmarchal ,
thanks for your reply. You are right about the fact that I have multiple versions of python on my system. That's why I used Conda to try to isolate things. Is there a better approach?

Following your suggestion, I tried without setting PYTHONPATH in my environment variables. 

> (SOFA) c:\users\frickers>echo %PYTHONPATH%
> %PYTHONPATH%

Unfortunately, the result is the same. When I try to include the SofaPython3 plugin I get the error about encodings.

Note that when I simply launch python in my Conda environment, it works fine:

> (SOFA) c:\users\frickers>python
> Python 3.10.15 | packaged by Anaconda, Inc. | (main, Oct  3 2024, 07:22:19) [MSC v.1929 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.

I asked a colleague to try installing and launching runSofa with the SofaPython3 plugin and he obtained the same error. So it's not just me.

**Reply by bakpaul** [reply]
Hello ! 

Looking around in forums, this error might be due to incompatibility between packages, I've read about some outdated packages that could mess with that or even numpy versions... 

Before going head down into this, could you try one thing ? libSofaPython3 is dynamically linked, which means that the library it will link upon is decided by the OS that uses it. I don't suspect anything but we might find some interesting results if you could run this software https://github.com/lucasg/Dependencies on the libSofaPython3 dll inside of your conda environment to see if it actually linked upon the conda packages or something unrelated. 

Please tell me how it goes.

**Reply by bakpaul** [reply]
If you cannot use this tool, it is just an open-source version of [dependencywalker](https://www.dependencywalker.com/)

**Reply by sebastienfricker** [reply]
Thank you Paul fpr the suggestion.
I ran DepencyWalker in Gui version, here is a screenshot:
![image](https://github.com/user-attachments/assets/88b11a05-1e8a-4538-a604-a4d1a6adaf86)
It seems to me that all DLL dependencies are properly loaded, but let me know if you see something suspicious or need additional output.

**Reply by sebastienfricker** [reply, marked as answer]
UPDATE: after struggling to add SofaPython3 in runSofa GUI, I tried to run it directly via a python script (in Spyder, in conda environment).
I set the PATH variable directly in the script via the following lines:

```
import os
import sys
 
#Create temporary environment variables, update PATH
os.environ['PYTHON_ROOT'] = r"C:\Users\FRICKERS\AppData\Local\anaconda3\envs\SOFA"
os.environ['PYTHONPATH'] = r"C:\Users\FRICKERS\SOFA\v24.06.00\plugins\SofaPython3\lib\python3\site-packages"
sys.path.insert(0, os.path.join(os.environ['PYTHON_ROOT'], "Lib"))
sys.path.insert(0, os.path.join(os.environ['PYTHON_ROOT'], "DLLs"))
sys.path.insert(0, os.environ['PYTHONPATH'])
```

After this header, the import of Sofa succeeds. I'm able to run a simulation and launch the Gui viewer. So at least I have a working solution!
